kratos/utilities/data_type_traits.h

```diff
@@ -0,0 +1,642 @@
+    {
+        if (rContainer.size() != 0) {
+            const auto stride = DataTypeTraits<ValueType>::Size(rContainer[0]);
+            for (unsigned int i = 0; i < rContainer.size(); ++i) {
+                DataTypeTraits<ValueType>::FillToContiguousData(pContiguousDataBegin + i * stride, rContainer[i]);
+            }
+        }
+    }
+
+    inline static void FillFromContiguousData(
+        ContainerType& rContainer,
+        PrimitiveType const * pContiguousDataBegin)
+    {
+        if (rContainer.size() != 0) {
+            const auto stride = DataTypeTraits<ValueType>::Size(rContainer[0]);
+            for (unsigned int i = 0; i < rContainer.size(); ++i) {
+                DataTypeTraits<ValueType>::FillFromContiguousData(rContainer[i], pContiguousDataBegin + i * stride);
+            }
+        }
+    }
+
+    ///@}
+};
+
+template<class TDataType>
+class DataBuffer
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using BufferedDataTypeTrait = DataTypeTraits<TDataType>;
+
+    using PrimitiveType = typename BufferedDataTypeTrait::PrimitiveType;
+
+    static constexpr bool HasContiguousPrimitiveData = BufferedDataTypeTrait::HasContiguousPrimitiveData;
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    inline PrimitiveType const * GetData(const TDataType& rContainer)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return BufferedDataTypeTrait::GetContiguousData(rContainer);
+        } else {
+            const auto size = BufferedDataTypeTrait::Size(rContainer);
+            if (mData.size() != size) {
+                mData.resize(size);
+            }
+            BufferedDataTypeTrait::FillToContiguousData(mData.data(), rContainer);
+            return mData.data();
+        }
+    }
+
+    inline PrimitiveType * GetData(TDataType& rContainer)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return BufferedDataTypeTrait::GetContiguousData(rContainer);
+        } else {
+            const auto size = BufferedDataTypeTrait::Size(rContainer);
+            if (mData.size() != size) {
+                mData.resize(size);
+            }
+            BufferedDataTypeTrait::FillToContiguousData(mData.data(), rContainer);
+            return mData.data();
+        }
+    }
+
+    void UpdateValues(TDataType& rContainer) const
+    {
+        if constexpr(!HasContiguousPrimitiveData) {
+            KRATOS_ERROR_IF_NOT(BufferedDataTypeTrait::Size(rContainer) == static_cast<unsigned int>(mData.size()))
+                << "Size mismatch [ rContainer flat size = " << BufferedDataTypeTrait::Size(rContainer)
+                << ", buffered data size = " << mData.size() << " ].\n";
+            BufferedDataTypeTrait::FillFromContiguousData(rContainer, mData.data());
+        }
+    }
+
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    std::conditional_t<HasContiguousPrimitiveData, void *, std::vector<PrimitiveType>> mData;
```

Comment:
functions in a *traits* class are weird enough, but doing buffering in it is on a whole new level of weirdness. I'd move this into its own class together with the functions that operate on them.

Reply:
done :)